### PR TITLE
[Feature]: Support deployment of panoptic segmentation models

### PR DESCRIPTION
**Pull request by RunningLeon**
## Motivation

Support the deployment of panoptic segmentation models like PanopticFPN, MaskFormer, Mask2Former on ONNX Runtime backend and TensorRT backends.

### Supported models

|                                                  Model                                                   |         Task          | OnnxRuntime | TensorRT | ncnn | PPLNN | OpenVINO |
| :------------------------------------------------------------------------------------------------------: | :-------------------: | :---------: | :------: | :--: | :---: | :------: |
|         [Panoptic FPN](https://github.com/open-mmlab/mmdetection/tree/main/configs/panoptic_fpn)         | Panoptic Segmentation |      Y      |    Y     |  N   |   N   |    N     |
|           [MaskFormer](https://github.com/open-mmlab/mmdetection/tree/main/configs/maskformer)           | Panoptic Segmentation |      Y      |    Y     |  N   |   N   |    N     |
| [Mask2Former](https://github.com/open-mmlab/mmdetection/tree/main/configs/mask2former)[\*](#mask2former) | Panoptic Segmentation |      Y      |    Y     |  N   |   N   |    N     |

### Reminder

- <i id="mask2former">Mask2Former</i> should use `TensorRT>=8.6.1` for dynamic shape inference.
-  Should use mmdet after this PR: https://github.com/open-mmlab/mmdetection/pull/10782

## BC-breaking (Optional)

None 

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by daigo0927** [comment]
> @daigo0927 hi, your commits fomr this PR #649 will be included in this PR, could you kindly sign the [CLA](https://cla-assistant.io/open-mmlab/mmdeploy?pullRequest=2347)? Thanks.

Thank you for the notification and I have signed the CLA. I'm really happy to hear the update 🎉

File: configs/mmdet/_base_/base_panoptic-seg_static.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+_base_ = ['../../_base_/onnx_config.py']
+
+codebase_config = dict(
+    type='mmdet',
+    task='ObjectDetection',
+    model_type='panoptic_end2end',
+    post_processing=dict(
+        export_postprocess_mask=False,
+        score_threshold=0.0,
+        confidence_threshold=0.005,  # for YOLOv3
```
Comment (AllentDan) [inline configs/mmdet/_base_/base_panoptic-seg_static.py:10]: `# for YOLOv3`？

File: mmdeploy/codebase/mmdet/models/detectors/panoptic_two_stage_segmentor.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            Defaults to True.
+
+    Returns:
+        tuple[Tensor, Tensor, Tensor, Tensor]:
+            (bboxes, labels, masks, semseg), `bboxes` of shape [N, num_det, 5],
+            `labels` of shape [N, num_det], `masks` of shape [N, roi_H, roi_W],
+            `semseg` of shape [N, num_sem_class, sem_H, sem_W].
+    """
+    ctx = FUNCTION_REWRITER.get_context()
+    data_samples = copy.deepcopy(data_samples)
```
Comment (AllentDan) [inline mmdeploy/codebase/mmdet/models/detectors/panoptic_two_stage_segmentor.py:37]: Got a RuntimeError: NYI: Named tensors are not supported with the tracer. I used PyTorch 1.12

**Review by AllentDan — COMMENTED** [review]
Got two errors:
1. batch_dets, batch_labels, batch_masks = outputs[:3]
ValueError: not enough values to unpack (expected 3, got 2)
2. [TRT] [E] 1: [defaultAllocator.cpp::deallocate::42] Error Code 1: Cuda Runtime (an illegal memory access was encountered)
panoptic_fpn with TensorRT 8.4.1.5

**Comment by AllentDan** [comment]
I used the following commands:
```shell
python tools/deploy.py configs/mmdet/panoptic-seg/panoptic-seg_maskformer_tensorrt_dynamic-320x512-1344x1344.py ../mmdetection/configs/panoptic_fpn/panoptic-fpn_r50_fpn_1x_coco.py https://download.openmmlab.com/mmdetection/v2.0/panoptic_fpn/panoptic_fpn_r50_fpn_1x_coco/panoptic_fpn_r50_fpn_1x_coco_20210821_101153-9668fd13.pth ../mmdetection/demo/demo.jpg --work-dir work_dirs/tensorrt/det/panoptic_fpn --device cuda --show --dump-info
```

**Comment by RunningLeon** [comment]
closes https://github.com/open-mmlab/mmdeploy/issues/298 https://github.com/open-mmlab/mmdeploy/issues/1303 https://github.com/open-mmlab/mmdeploy/issues/1880

**Comment by RunningLeon** [comment]
> I used the following commands:
> 
> ```shell
> python tools/deploy.py configs/mmdet/panoptic-seg/panoptic-seg_maskformer_tensorrt_dynamic-320x512-1344x1344.py ../mmdetection/configs/panoptic_fpn/panoptic-fpn_r50_fpn_1x_coco.py https://download.openmmlab.com/mmdetection/v2.0/panoptic_fpn/panoptic_fpn_r50_fpn_1x_coco/panoptic_fpn_r50_fpn_1x_coco_20210821_101153-9668fd13.pth ../mmdetection/demo/demo.jpg --work-dir work_dirs/tensorrt/det/panoptic_fpn --device cuda --show --dump-info
> ```

for panoptic-fpn model, should use `configs/mmdet/panoptic-seg/panoptic-seg_panoptic-fpn_xxx.py`

**Review by AllentDan — APPROVED** [review]
LGTM

File: mmdeploy/apis/onnx/export.py [hunk 3]
```diff
@@ -127,7 +127,14 @@ def wrapper(*arg, **kwargs):
             patched_model.forward = wrap_forward(patched_model.forward)
             patched_model.forward = partial(patched_model.forward,
                                             **input_metas)
-
+        # force to export on cpu
```
Comment (grimoire) [inline mmdeploy/apis/onnx/export.py:130]: What if here are op that only support gpu?

Comment (RunningLeon) [reply mmdeploy/apis/onnx/export.py:130]: maybe we could add a dummy custom op to make it runnable on cpu.

**Review by grimoire — APPROVED** [review]

**Comment by taifyang** [comment]
Does mmdeploy SDK support MaskFormer, Mask2Former inference？

**Comment by taifyang** [comment]
Does mmdeploy SDK support MaskFormer, Mask2Former panoptic-seg for onnx and trt？

**Comment by taifyang** [comment]
> closes #298 #1303 #1880

Will mmdeploy SDK support MaskFormer, Mask2Former panoptic-seg for onnx and trt? I convert pth to onnx successfully but sdk inference failed.

**Comment by taifyang** [comment]
> closes #298 #1303 #1880

Will mmdeploy SDK support MaskFormer, Mask2Former panoptic-seg for onnx and trt? I convert pth to onnx successfully but sdk inference failed. @RunningLeon

**Comment by RunningLeon** [comment]
> > closes #298 #1303 #1880
> 
> Will mmdeploy SDK support MaskFormer, Mask2Former panoptic-seg for onnx and trt? I convert pth to onnx successfully but sdk inference failed. @RunningLeon

@taifyang hi, sorry for the late reply. SDK has not supported for mask2former series. I'm afraid you have to implement it by yourself. Thanks for your understanding.

**Comment by SparKgod1** [comment]
@RunningLeon Hello, I have modified the original output of Mask2Former, which consists of cls_logits and mask_logits, by rewriting the forward function of the fusion head to produce outputs that can be processed by a C++ SDK. The outputs are now dets, labels, and masks. However, I encountered an issue where the SDK runs into the function located at mmdeploy/mmdeploy/csrc/mmdeploy/codebase/mmdet/object_detection.cpp, and it seems that this function cannot handle the masks output, resulting in an empty mask in the SDK output.
I noticed that mmdeploy/mmdeploy/csrc/mmdeploy/codebase/mmdet/instance_segmentation.cpp appears to be capable of handling cases where the TensorRT model output includes masks. However, I am unsure how to modify the code to ensure that the output is processed by this function instead. I would greatly appreciate your guidance on how to make this adjustment. Your assistance would be extremely valuable to me. Thank you very much!

**Pull request merged**